File: sdk/python/src/axern_sdk/context.py
"""Explicit loader for Axern CLI contexts."""

from __future__ import annotations

import json
from dataclasses import dataclass
from pathlib import Path
from typing import Any


@dataclass(frozen=True)
class TLSContext:
    ca_cert: str
    cert: str
    key: str
    server_name: str = ""


@dataclass(frozen=True)
class AxernContext:
    name: str
    endpoint: str
    service_url: str
    ssh_endpoint: str
    ssh_identity_file: str
    tls: TLSContext
    proxy_mode: str


def load_context(path: str | Path, name: str = "") -> AxernContext:
    """Load one named context without implicitly reading the user directory."""

    config_path = Path(path).expanduser()
    raw = json.loads(config_path.read_text(encoding="utf-8"))
    if not isinstance(raw, dict):
        raise ValueError("Axern config must be an object")
    _reject_unknown(raw, {"current_context", "contexts", "agent_profiles"}, "config")
    context_name = name or _string(raw, "current_context")
    if not context_name:
        raise ValueError("Axern context name is required")
    contexts = raw.get("contexts")
    if not isinstance(contexts, dict) or context_name not in contexts:
        raise ValueError(f"Axern context {context_name!r} was not found")
    context = contexts[context_name]
    if not isinstance(context, dict):
        raise ValueError(f"Axern context {context_name!r} must be an object")
    _reject_unknown(context, {"endpoint", "service_url", "ssh_endpoint", "ssh_identity_file", "tls", "proxy_mode"}, "context")
    tls = context.get("tls")
    if not isinstance(tls, dict):
        raise ValueError("context.tls must be an object")
    _reject_unknown(tls, {"ca_cert", "cert", "key", "server_name"}, "context.tls")
    endpoint = _string(context, "endpoint")
    ca_cert = _string(tls, "ca_cert")
    cert = _string(tls, "cert")
    key = _string(tls, "key")
    if not endpoint or not ca_cert or not cert or not key:
        raise ValueError("context requires endpoint and tls.ca_cert, tls.cert, and tls.key")
    proxy_mode = _string(context, "proxy_mode") or "env"
    if proxy_mode not in {"env", "direct"}:
        raise ValueError("context.proxy_mode must be 'env' or 'direct'")
    return AxernContext(
        name=context_name,
        endpoint=endpoint,
        service_url=_string(context, "service_url"),
        ssh_endpoint=_string(context, "ssh_endpoint"),
        ssh_identity_file=_string(context, "ssh_identity_file"),
        tls=TLSContext(ca_cert=ca_cert, cert=cert, key=key, server_name=_string(tls, "server_name")),
        proxy_mode=proxy_mode,
    )


def _string(value: dict[str, Any], key: str) -> str:
    item = value.get(key, "")
    if not isinstance(item, str):
        raise ValueError(f"{key} must be a string")
    return item.strip()


def _reject_unknown(value: dict[str, Any], allowed: set[str], path: str) -> None:
    unknown = sorted(set(value) - allowed)
    if unknown:
        raise ValueError(f"{path} contains unknown field {unknown[0]!r}")

File: sdk/go/clientconfig/context.go
// Package clientconfig loads the explicit context file shared by the Axern CLI and SDKs.
package clientconfig

import (
	"bytes"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"os"
	"path/filepath"
	"sort"
	"strings"
)

const (
	ProxyModeEnv    = "env"
	ProxyModeDirect = "direct"
)

type TLS struct {
	CACert     string `json:"ca_cert,omitempty"`
	Cert       string `json:"cert,omitempty"`
	Key        string `json:"key,omitempty"`
	ServerName string `json:"server_name,omitempty"`
}

type Context struct {
	Endpoint        string `json:"endpoint"`
	ServiceURL      string `json:"service_url,omitempty"`
	SSHEndpoint     string `json:"ssh_endpoint,omitempty"`
	SSHIdentityFile string `json:"ssh_identity_file,omitempty"`
	TLS             TLS    `json:"tls"`
	ProxyMode       string `json:"proxy_mode,omitempty"`
}

type File struct {
	CurrentContext string              `json:"current_context,omitempty"`
	Contexts       map[string]*Context `json:"contexts,omitempty"`
	AgentProfiles  json.RawMessage     `json:"agent_profiles,omitempty"`
}

func DefaultPath() string {
	home, err := os.UserHomeDir()
	if err != nil || home == "" {
		return "config.json"
	}
	return filepath.Join(home, ".config", "axern", "config.json")
}

func Load(path string) (*File, error) {
	if path == "" {
		path = DefaultPath()
	}
	data, err := os.ReadFile(path)
	if errors.Is(err, os.ErrNotExist) {
		return &File{Contexts: map[string]*Context{}}, nil
	}
	if err != nil {
		return nil, err
	}
	var cfg File
	decoder := json.NewDecoder(bytes.NewReader(data))
	decoder.DisallowUnknownFields()
	if err := decoder.Decode(&cfg); err != nil {
		return nil, fmt.Errorf("parse axern config %q: %w", path, err)
	}
	if err := decoder.Decode(&struct{}{}); !errors.Is(err, io.EOF) {
		return nil, fmt.Errorf("parse axern config %q: multiple JSON values", path)
	}
	if cfg.Contexts == nil {
		cfg.Contexts = map[string]*Context{}
	}
	for name, context := range cfg.Contexts {
		if err := Validate(context); err != nil {
			return nil, fmt.Errorf("invalid axern context %q: %w", name, err)
		}
	}
	return &cfg, nil
}

func Resolve(path, name string) (string, *Context, bool, error) {
	cfg, err := Load(path)
	if err != nil {
		return "", nil, false, err
	}
	if name == "" {
		name = cfg.CurrentContext
	}
	if name == "" {
		return "", nil, false, nil
	}
	context, ok := cfg.Contexts[name]
	if !ok || context == nil {
		return "", nil, false, fmt.Errorf("axern context %q not found in %s", name, resolvedPath(path))
	}
	return name, context, true, nil
}

func Validate(context *Context) error {
	if context == nil {
		return fmt.Errorf("context is required")
	}
	if strings.TrimSpace(context.Endpoint) == "" {
		return fmt.Errorf("endpoint is required")
	}
	if strings.TrimSpace(context.TLS.CACert) == "" || strings.TrimSpace(context.TLS.Cert) == "" || strings.TrimSpace(context.TLS.Key) == "" {
		return fmt.Errorf("tls.ca_cert, tls.cert, and tls.key are required")
	}
	switch strings.TrimSpace(context.ProxyMode) {
	case "", ProxyModeEnv, ProxyModeDirect:
		return nil
	default:
		return fmt.Errorf("proxy_mode must be %q or %q", ProxyModeEnv, ProxyModeDirect)
	}
}

func ContextNames(cfg *File) []string {
	if cfg == nil {
		return nil
	}
	names := make([]string, 0, len(cfg.Contexts))
	for name := range cfg.Contexts {
		names = append(names, name)
	}
	sort.Strings(names)
	return names
}

func resolvedPath(path string) string {
	if path != "" {
		return path
	}
	return DefaultPath()
}
